synapse/rest/client/v1/room.py

```diff
@@ -266,6 +266,285 @@ def on_PUT(self, request, room_id, event_type, txn_id):
         )
 
 
+class RoomBatchSendEventRestServlet(TransactionRestServlet):
+    """
+    API endpoint which can insert a chunk of events historically back in time
+    next to the given `prev_event`.
+
+    `chunk_id` comes from `next_chunk_id `in the response of the batch send
+    endpoint and is derived from the "insertion" events added to each chunk.
+    It's not required for the first batch send.
+
+    `state_events_at_start` is used to define the historical state events
+    needed to auth the events like join events. These events will float
+    outside of the normal DAG as outlier's and won't be visible in the chat
+    history which also allows us to insert multiple chunks without having a bunch
+    of `@mxid joined the room` noise between each chunk.
+
+    `events` is chronological chunk/list of events you want to insert.
+    There is a reverse-chronological constraint on chunks so once you insert
+    some messages, you can only insert older ones after that.
+    tldr; Insert chunks from your most recent history -> oldest history.
+
+    POST /_matrix/client/r0/rooms/<roomID>/batchsend?prev_event=<eventID>&chunk_id=<chunkID>
+    {
+        "events": [ ... ],
+        "state_events_at_start": [ ... ]
+    }
+    """
+
+    def __init__(self, hs):
+        super().__init__(hs)
+        self.hs = hs
+        self.store = hs.get_datastore()
+        self.state_store = hs.get_storage().state
+        self.event_creation_handler = hs.get_event_creation_handler()
+        self.room_member_handler = hs.get_room_member_handler()
+        self.auth = hs.get_auth()
+
+    def register(self, http_server):
+        # /rooms/$roomid/batchsend
+        PATTERNS = "/rooms/(?P<room_id>[^/]*)/batchsend"
+        register_txn_path(self, PATTERNS, http_server, with_get=True)
```

Comment:
Unfortunately `register_txn_path` adds the apis to `/_matrix/client/r0` and `api/v1`, so we'll need to change it to allow only adding to the unstable prefix (which I think is just passing through some options).

We'll also want to prefix the paths with `/org.matrix.msc2716/`, so that the full path is `/_matrix/client/unstable/org.matrix.msc2716/rooms/...`

Sorry for not spotting this previously.

Reply:
It looks there are some easy patterns for this already with the `RestServlet` 🍀

 - [`synapse/rest/client/v1/room.py#L913-L919`](https://github.com/matrix-org/synapse/blob/18edc9ab06d8ed07c1cac918057226fad18030ce/synapse/rest/client/v1/room.py#L913-L919)
 - [`synapse/rest/client/v1/room.py#L1004-L1010`](https://github.com/matrix-org/synapse/blob/18edc9ab06d8ed07c1cac918057226fad18030ce/synapse/rest/client/v1/room.py#L1004-L1010)